crates/fluvio-test/src/tests/producer_fail/mod.rs

```diff
@@ -0,0 +1,101 @@
+use std::any::Any;
+use std::sync::Arc;
+
+use fluvio::{RecordKey, TopicProducer, FluvioAdmin, FluvioError};
+use fluvio_controlplane_metadata::partition::PartitionSpec;
+use structopt::StructOpt;
+
+use fluvio_test_derive::fluvio_test;
+use fluvio_test_util::test_meta::derive_attr::TestRequirements;
+use fluvio_test_util::test_meta::environment::EnvironmentSetup;
+use fluvio_test_util::test_meta::{TestOption, TestCase};
+use fluvio_test_util::test_meta::test_result::TestResult;
+use fluvio_test_util::test_runner::test_driver::{TestDriver};
+use fluvio_test_util::test_runner::test_meta::FluvioTestMeta;
+
+#[derive(Debug, Clone)]
+pub struct ProducerBatchTestCase {
+    pub environment: EnvironmentSetup,
+    pub option: ProduceBatchTestOption,
+}
+
+impl From<TestCase> for ProducerBatchTestCase {
+    fn from(test_case: TestCase) -> Self {
+        let producer_option = test_case
+            .option
+            .as_any()
+            .downcast_ref::<ProduceBatchTestOption>()
+            .expect("ProducerBatchTestOption")
+            .to_owned();
+        Self {
+            environment: test_case.environment,
+            option: producer_option,
+        }
+    }
+}
+
+#[derive(Debug, Clone, StructOpt, Default, PartialEq)]
+#[structopt(name = "Fluvio Producer Batch Test")]
+pub struct ProduceBatchTestOption {}
+
+impl TestOption for ProduceBatchTestOption {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+}
+
+#[fluvio_test(topic = "test")]
+pub async fn produce_batch(
+    mut test_driver: Arc<FluvioTestDriver>,
+    mut test_case: TestCase,
+) -> TestResult {
+    println!("Starting produce_batch test");
+
+    let topic_name = test_case.environment.topic_name();
+    let producer: TopicProducer = test_driver.create_producer(&topic_name).await;
+
+    println!("Created producer");
+
+    let leader = {
+        let admin: FluvioAdmin = test_driver.client().admin().await;
+        let partitions = admin
+            .list::<PartitionSpec, _>(vec![])
+            .await
+            .expect("partitions");
+        let test_topic = &partitions[0];
+        let leader = test_topic.spec.leader;
+        leader
+    };
+
+    println!("Found leader {}", leader);
+
+    let cluster_manager = test_driver
+        .get_cluster()
+        .expect("cluster")
+        .env_driver()
+        .create_cluster_manager();
+
+    println!("Got cluster manager");
+
+    let result: Result<_, FluvioError> = (|| async move {
+        for i in 0..1_000 {
+            producer.send(RecordKey::NULL, i.to_string()).await?;
+        }
+        println!("Send 1000");
+
+        cluster_manager.terminate_spu(leader).expect("terminate");
+        println!("Terminate SPU");
+
+        for i in 0..1_000 {
+            producer.send(RecordKey::NULL, i.to_string()).await?;
```

Comment:
need assertion here

Reply:
So I can add an assertion in the first loop (before SPU is terminated), but I can't add an assertion here. The reason is that, ever since the SPU died, we don't know _which_ call to `send` is going to fail, because we don't know when the dispatcher is going to try to flush.

The way this test is set up is that each call to `send` uses `?` so that when an Err crops up _somewhere_, it will get caught by the closure surrounding it. If you look at the end of the test, there is an assertion on the final `Result` that comes out of the closure. Meaning that we are asserting that at least one call to `send` or `flush` failed, but we don't necessarily know which one.